datasets/preprocess_utils.py

```diff
@@ -243,3 +243,181 @@ def bin_stream_hilmi24(
+    mGDES += fudgeShift # Peak shift
+    return mGDES
+
+def get_decam_g_r_i_z(iso_name = 'Dotter', iso_age = 11.5, iso_metallicity = 0.00016, iso_distance_modulus = 16.807, iso_stellar_mass = 2e4):
+    """Get simulated DECam g, r, i, z magnitudes from the AAU isochrone"""
+    
+    # Simulated AAU isochrone for g and r bands 
+    iso_g_r = isochrone.factory(name=iso_name,
+                            age=iso_age,  # Gyr
+                            metallicity=iso_metallicity,  # Approximate Z value for the stream metallicity
+                            distance_modulus=iso_distance_modulus,  # Average distance modulus from the stream properties
+                            )
+
+    # Simulated AAU isochrone for i and z bands
+    iso_i_z = isochrone.factory(name=iso_name,
+                            age=iso_age,  # Gyr
+                            metallicity=iso_metallicity,  # Approximate Z value for the stream metallicity
+                            distance_modulus=iso_distance_modulus,  # Average distance modulus from the stream properties
+                            band_1 = 'i',
+                            band_2 = 'z'
+                            )
+
+    decam_g, decam_r = iso_g_r.simulate(stellar_mass=iso_stellar_mass)
+    decam_i, decam_z = iso_i_z.simulate(stellar_mass=iso_stellar_mass)
+
+    return decam_g, decam_r, decam_i, decam_z
+
+def get_gaia_g(decam_g, decam_r, decam_i, decam_z):
+    """Get simulated Gaia G band magnitudes from the AAU isochrone"""
+
+    gaia_g = gaia_transform(decam_g, decam_r, decam_i, decam_z, version='edr3')
+
+    filtered_gaia_g = []
+    
+    color = decam_g - decam_r + 0.04
+    min_color = 0.2
+    max_color = max(color)
+
+    min_mag = 14.875
+    max_mag = 19.512
+
+    for c, m, j in zip(color, decam_r, gaia_g):
+            if min_color <= c <= max_color and min_mag <= m <= max_mag:
+                    filtered_gaia_g.append(j)
+    return filtered_gaia_g
+
+def get_pm_uncertainties(gaia_g):
+    """Get proper motion uncertainties for the simulated stars"""
+    
+    # Calculate proper motion uncertainties for Gaia DR3
+    pmra_unc, pmdec_unc = proper_motion_uncertainty(gaia_g, release='dr3')
+
+    pmra_err = pmra_unc/1000
+    pmdec_err = pmdec_unc/1000
+        
+    return pmra_err, pmdec_err
+
+def add_uncertainty(
+    feat: np.ndarray,
+    dist = True,
+    v_r = True,
+    pm_phi1 = True,
+    pm_phi2 = True):
+    """ Add uncertainties to the features: distances, radial velocities, proper motions in phi1,
+    and proper motions in phi2. """
+    
+    
+    if pm_phi1 or pm_phi2:
+        decam_g, decam_r, decam_i, decam_z = get_decam_g_r_i_z()
+        gaia_g = get_gaia_g(decam_g, decam_r, decam_i, decam_z)
+        pmra_err, pmdec_err = get_pm_uncertainties(np.array(gaia_g))
+    
+    if pm_phi1:
+        for i in range(len(feat[:, 1])):
+            feat[i, 1] = np.random.normal(loc=feat[i, 1], scale=np.random.choice(pmra_err), size=1)[0]
```

Comment:
While this will produce the correct result, this is not a very inefficient way to do this. 

In Python, for loop is extremely efficient. NumPy, which is C in the backend, use vectorized and array operation to speed things up significantly. When using NumPy, use array operation unless you absolutely don't have a way to. I suggest reading this bit and then go back and rewrite these functions. 

https://dev.to/chamodperera/replacing-for-loops-with-vectorization-in-python-21m6?utm_source=chatgpt.com

Reply:
fixed